src/fighter_state.rs

```diff
@@ -554,10 +599,111 @@ fn attacking(
+) {
+    for (
+        entity,
+        mut animation,
+        mut velocity,
+        facing,
+        stats,
+        meta_handle,
+        mut punching,
+        player,
+        enemy,
+    ) in &mut fighters
+    {
+        let is_player = player.is_some();
+        let is_enemy = enemy.is_some();
+        if !is_player && !is_enemy {
+            // This system only knows how to attack for players and enemies
+            continue;
+        }
+
+        if !punching.has_started {
+            punching.has_started = true;
+
+            // Start the attack  from the beginning
+            animation.play(Punching::ANIMATION, false);
+
+            if let Some(fighter) = fighter_assets.get(meta_handle) {
+                let mut offset = fighter.attack.hitbox_offset;
+                if facing.is_left() {
+                    offset *= -1.0
+                }
+                let attack_frames = fighter.attack.frames;
+                // Spawn the attack entity
+                let attack_entity = commands
+                    // .spawn_bundle(TransformBundle::default())
+                    .spawn_bundle(TransformBundle::from_transform(
+                        Transform::from_translation(offset.extend(0.0)),
+                    ))
+                    .insert(Sensor)
+                    .insert(ActiveEvents::COLLISION_EVENTS)
+                    .insert(ActiveCollisionTypes::default() | ActiveCollisionTypes::STATIC_STATIC)
+                    .insert(CollisionGroups::new(
+                        if is_player {
+                            BodyLayers::PLAYER_ATTACK
+                        } else {
+                            BodyLayers::ENEMY_ATTACK
+                        },
+                        if is_player {
+                            BodyLayers::ENEMY
+                        } else {
+                            BodyLayers::PLAYER
+                        },
+                    ))
+                    .insert(Attack {
+                        damage: stats.damage,
+                        velocity: if facing.is_left() {
+                            Vec2::NEG_X
+                        } else {
+                            Vec2::X
+                        } * Vec2::new(consts::ATTACK_VELOCITY, 0.0),
+                    })
+                    .insert(attack_frames)
+                    .id();
+                commands.entity(entity).push_children(&[attack_entity]);
+
+                // Play attack sound effect
+                if let Some(effects) = fighter.audio.effect_handles.get(Punching::ANIMATION) {
+                    let fx_playback = AnimationAudioPlayback::new(
+                        Punching::ANIMATION.to_owned(),
+                        effects.clone(),
+                    );
+                    commands.entity(entity).insert(fx_playback);
+                }
+            } else {
+                //what should we actually do if we cant read attack metadata from fighter?
```

Comment:
Technically this shouldn't happen because we have logic to make sure all assets are loaded before starting gameplay. We could maybe panic with a message saying that much so that we know if we accidentally break pre-loading assets maybe?

Reply:
So this is the correct way to query the metadata then? Shouldn't have to worry about the handle not returning something?

Reply:
Yeah. The handle should always return or we've done something wrong in the loader.

So we could either `.expect("Handle not loaded")` or something like that, or we could `warn!("Skipping attack due to un-loaded fighter meta")`.

I'm on the fence between the two. If we panic, we won't accidentally mis the log message, but it might crash the game for a user who doesn't know what to do about it, and it might only cause a 1-frame problem anyway and it could have been safely ignored.

If we don't panic, we face the possibility of not seeing the warning during development, and being more likely to ship a release of the game with a bug in it.

Difficult trade-off to decide between. :man_shrugging: